File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.cpp
#include <string>
#include <vector>
#include <queue>
#include <map>
using namespace std;

map<int,int> ma;
vector<int> solution(vector<string> operations) {
    vector<int> answer;
    priority_queue<int,vector<int>,greater<int>> minq;
    priority_queue<int> maxq;
    for(string op:operations){
        char o=op[0];
        int d=stoi(op.substr(2));
        if(o=='I'){
            minq.push(d);
            maxq.push(d);
            ma[d]++;
        }else{
            if(d==1){
                while(!maxq.empty()&&ma[maxq.top()]==0){
                    maxq.pop();
                }
                if(!maxq.empty()){
                    ma[maxq.top()]--;
                    maxq.pop();
                }
            }else{
                while(!minq.empty()&&ma[minq.top()]==0){
                    minq.pop();
                }
                if(!minq.empty()){
                    ma[minq.top()]--;
                    minq.pop();
                }
            }
        }
    }
    while(!minq.empty()&&ma[minq.top()]==0){
        minq.pop();
    }
    while(!maxq.empty()&&ma[maxq.top()]==0){
        maxq.pop();
    }
    if(minq.empty()||maxq.empty()){
        return {0,0};
    }
    return {maxq.top(),minq.top()};
}

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;

class Solution {
	public int[] solution(String[] operations) {
		int[] answer = {};

		Map<Integer, Integer> map = new HashMap<>();
		Queue<Integer> maxQ = new PriorityQueue<>((n1, n2) -> n2 - n1);
		Queue<Integer> minQ = new PriorityQueue<>();

		for (String op: operations) {
			char i = op.charAt(0);
			int num = Integer.parseInt(op.substring(2));
			if (i == 'I') {
				minQ.add(num);
				maxQ.add(num);
				map.put(num, map.getOrDefault(num, 0) + 1);
			} else {
				if (num == 1) {
					// ma[maxQ.top()] == 0
					while (!maxQ.isEmpty() && map.getOrDefault(maxQ.peek(), 0) == 0) {
						maxQ.poll();
					}
					if (!maxQ.isEmpty()) {
						// ma[maxQ.top()]--;
						map.put(maxQ.peek(), map.get(maxQ.peek()) - 1);
						maxQ.poll();
					}
				} else {
					while (!minQ.isEmpty() && map.getOrDefault(minQ.peek(), 0) == 0) {
						minQ.poll();
					}
					if (!minQ.isEmpty()) {
						map.put(minQ.peek(), map.get(minQ.peek()) - 1);
						minQ.poll();
					}
				}
			}
		}
		while (!maxQ.isEmpty() && map.getOrDefault(maxQ.peek(), 0) == 0) {
			maxQ.poll();
		}
		while (!minQ.isEmpty() && map.getOrDefault(minQ.peek(), 0) == 0) {
			minQ.poll();
		}
		if (minQ.isEmpty() || maxQ.isEmpty()) {
			return new int[]{0, 0};
		}

		return new int[]{maxQ.peek(), minQ.peek()};
	}
}